web/source/osk/visualKeyboard.ts

```diff
@@ -775,24 +779,38 @@ namespace com.keyman.osk {
             keys[j]['widthpc']=keyPercent;
             padPercent = Math.round(keys[j]['padpc'] * objectWidth); // Math.round(parseInt(keys[j]['pad'],10)*objectWidth/totalWidth);
             keys[j]['padpc']=padPercent;
+
+            // recompute center's x-coord
+            (<ActiveKey> keys[j]).proportionalX = (totalPercent + padPercent + (keyPercent/2))/objectWidth;
+            (<ActiveKey> keys[j]).proportionalWidth = keyPercent / objectWidth;
+
             totalPercent += padPercent+keyPercent;
           }
 
           // Allow for right OSK margin (15 layout units)
-          totalPercent += Math.round(15*objectWidth/row.totalWidth);
+          let rightMargin = Math.round(15*objectWidth/layer.totalWidth);
+          totalPercent += rightMargin;
 
           // If a single key, and padding is negative, add padding to right align the key
           if(keys.length == 1 && parseInt(keys[0]['pad'],10) < 0) {
             keyPercent = Math.round(keys[0]['widthpc'] * objectWidth); //Math.round(parseInt(keys[0]['width'],10)*objectWidth/totalWidth);
             keys[0]['widthpc']=keyPercent;
             totalPercent += keyPercent;
             keys[0]['padpc']=(objectWidth-totalPercent);
+
+            // recompute center's x-coord
+            (<ActiveKey> keys[0]).proportionalX = (totalPercent - rightMargin - keyPercent/2)/objectWidth;
+            (<ActiveKey> keys[0]).proportionalWidth = keyPercent / objectWidth;
           } else if(keys.length > 0) {
             j=keys.length-1;
             padPercent = Math.round(keys[j]['padpc'] * objectWidth); //Math.round(parseInt(keys[j]['pad'],10)*objectWidth/totalWidth);
             keys[j]['padpc']=padPercent;
             totalPercent += padPercent;
-            keys[j]['widthpc']=(objectWidth-totalPercent);
+            keys[j]['widthpc']= keyPercent = (objectWidth-totalPercent);
+
+            // recompute center's x-coord
+            (<ActiveKey> keys[j]).proportionalX = (objectWidth - rightMargin - keyPercent/2)/objectWidth;
+            (<ActiveKey> keys[j]).proportionalWidth = keyPercent / objectWidth;
```

Comment:
More code duplication, may be made clearer by factoring out a helper function.

Reply:
About that... this _isn't_ code duplication.

One of our goals for v13 is to support a proper 'headless' mode for KMW that facilitates unit testing.  Such a mode won't have any use for DOM elements, so the near-equivalent in ActiveLayout.ts is designed to provide _default_ values for these properties.  I've made sure that the calculations get close to the values used when not headless, but it's not perfect.

Naturally, we want the calculations to be as accurate as possible.  When KMW is using the DOM, there are various tweaks that occur to the layout geometry, such as rounding the width of key elements - and those are dependent upon the active display properties of the browser.  To ensure that our fat-finger calculations are as precise as possible, similar **but different** calculations to get the _true_ values seen by the user are necessary - that's what you see here.

It might be possible to make something of a helper function to handle both cases, but that helper function itself wouldn't be very clear - there'd be quite a number of parameters involved to get everything just right.

Reply:
Ah, gotcha!